### PR TITLE
Editorial: replace Type(...) notation with a proper Type AO

### LINKED ISSUE #2870: make Type(...) macro an AO and return a spec enum
The usage of `Type(...)` in the spec is quite strange and doesn't follow our usual editorial conventions. I would prefer if we just made it a regular AO and had it return spec enums, which can be more comfortably passed around and compared than the built-in constructor or whatever it is that is returned today. Also, we won't have to think about realm weirdness.

### PR DESCRIPTION
**Pull request by michaelficarra**
Fixes #2870. This aligns with our goal of reducing our core concepts.

1. I believe this is the last place where we reified ECMAScript language types. 
1. Fewer special forms and macros, more AOs.

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -6415,7 +6437,7 @@ <h1>
         <dd>It is used to create a List value whose elements are provided by the indexed properties of _obj_. _elementTypes_ contains the names of ECMAScript Language Types that are allowed for element values of the List that is created.</dd>
       </dl>
       <emu-alg>
-        1. If _elementTypes_ is not present, set _elementTypes_ to « Undefined, Null, Boolean, String, Symbol, Number, BigInt, Object ».
+        1. If _elementTypes_ is not present, set _elementTypes_ to « ~undefined~, ~null~, ~boolean~, ~string~, ~symbol~, ~number~, ~bigint~, ~object~ ».
```
Comment (bakkot) [inline spec.html:6440]: Presumably the signature of this AO needs updating, or alternatively we could `<dfn>` "names of ECMAScript Language Types".

Comment (michaelficarra) [reply spec.html:6440]: I think we don't need this kind of generality anyway. `elementTypes` only ever has two values. Maybe we replace it with an enum?

Comment (bakkot) [reply spec.html:6440]: Temporal calls it with `« Object »`, though maybe that's one of the ones you're referring to?

I think making this more specific would actually make it _more_ verbose, so I'm inclined to stick with the general version, but not strongly.

File: spec.html [hunk 2]
```diff
@@ -1095,15 +1095,37 @@ <h1>Identity</h1>
...
 
   <emu-clause id="sec-ecmascript-language-types">
     <h1>ECMAScript Language Types</h1>
     <p>An <dfn variants="ECMAScript language types">ECMAScript language type</dfn> corresponds to values that are directly manipulated by an ECMAScript programmer using the ECMAScript language. The ECMAScript language types are Undefined, Null, Boolean, String, Symbol, Number, BigInt, and Object. An <dfn variants="ECMAScript language values">ECMAScript language value</dfn> is a value that is characterized by an ECMAScript language type.</p>
 
+    <emu-clause id="sec-type-ao" type="abstract operation">
+      <h1>
+        Type (
+          _v_: an ECMAScript language value,
+        ): one of ~undefined~, ~null~, ~boolean~, ~string~, ~symbol~, ~number~, ~bigint~, or ~object~
```
Comment (jmdyck) [inline spec.html:1110]:
```suggestion
        ): ~undefined~, ~null~, ~boolean~, ~string~, ~symbol~, ~number~, ~bigint~, or ~object~
```
(We don't use "one of" in this kind of context.)

File: spec.html [hunk 3]
```diff
@@ -1095,15 +1095,37 @@ <h1>Identity</h1>
...
   <p>Algorithms within this specification manipulate values each of which has an associated type. The possible value types are exactly those defined in this clause. Types are further subclassified into ECMAScript language types and specification types.</p>
-  <p>Within this specification, the notation “Type(_x_)” is used as shorthand for “the <em>type</em> of _x_” where “type” refers to the ECMAScript language and specification types defined in this clause.</p>
 
   <emu-clause id="sec-ecmascript-language-types">
     <h1>ECMAScript Language Types</h1>
     <p>An <dfn variants="ECMAScript language types">ECMAScript language type</dfn> corresponds to values that are directly manipulated by an ECMAScript programmer using the ECMAScript language. The ECMAScript language types are Undefined, Null, Boolean, String, Symbol, Number, BigInt, and Object. An <dfn variants="ECMAScript language values">ECMAScript language value</dfn> is a value that is characterized by an ECMAScript language type.</p>
 
+    <emu-clause id="sec-type-ao" type="abstract operation">
+      <h1>
+        Type (
```
Comment (jmdyck) [inline spec.html:1108]: I was thinking this AO would be better placed in section 7 "Abstract Operations". (Note that its first use occurs in 7.2.9.) Now I'm less sure, but I thought I'd mention it anyway.

Comment (michaelficarra) [reply spec.html:1108]: I think I like it here. Section 7 AOs should probably be dispersed to relevant sections.

**Comment by michaelficarra** [comment]
Feedback addressed.

**Comment by michaelficarra** [comment]
https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeinternal will need to be updated (and possibly others).

**Review by syg — APPROVED** [review]

**Comment by syg** [comment]
Recapping editor call discussion:

Let's hold off on this as @ljharb may have cycles to do the following to get us to the desired end-state of not having any type reflection at all:

1. Introduce a SameType() AO
2. Replace the CreateArrayFromList type list with a 2-state enum for all types and property keys only
3. Do something else for the BigInt dispatch table, either splitting 2 tables or using a difference phrase

**Comment by ljharb** [comment]
See #3408 and #3409 for 1 and 2.

**Comment by ljharb** [comment]
→ #3410 and #3420

**Pull request closed without merging**